support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,313 @@
+		}
+	}
+	return spiffeID
+}
+
+func (r *PodReconciler) makeID(pathFmt string, pathArgs ...interface{}) string {
+	id := url.URL{
+		Scheme: "spiffe",
+		Host:   r.TrustDomain,
+		Path:   path.Clean(fmt.Sprintf(pathFmt, pathArgs...)),
+	}
+	return id.String()
+}
+
+func (r *PodReconciler) makeParentIDForPod(pod *corev1.Pod) string {
+	nodeName := pod.Spec.NodeName
+	if nodeName == "" {
+		return ""
+	}
+	return fmt.Sprintf("%s/%s", r.RootID, nodeName)
+}
+
+func (r *PodReconciler) makeParentID(obj ObjectWithMetadata) string {
+	return r.makeParentIDForPod(obj.(*corev1.Pod))
+}
+
+func (r *PodReconciler) getSelectors(namespacedName types.NamespacedName) []*common.Selector {
+	return []*common.Selector{
+		r.k8sWorkloadSelector(PodNamespaceSelector, namespacedName.Namespace),
+		r.k8sWorkloadSelector(PodNameSelector, namespacedName.Name),
+	}
+}
+
+func (r *PodReconciler) getAllEntries(ctx context.Context) ([]*common.RegistrationEntry, error) {
+	// Parents for an entry are not guaranteed to exist. This means we cannot do a search by parent ID
+	// starting from rootId to find nodes, then find pods parented to those nodes. Instead we have to
+	// get the full set of entries, and scan them for parentIds that match the format we use for a
+	// node's ID. This is probably faster anyway: most entries in spire are going to be for pods, so we
+	// may as well just load the whole lot.
+	// TODO: Move to some kind of poll and cache and notify system, so multiple controllers don't have to poll.
+	allEntries, err := r.SpireClient.FetchEntries(ctx, &common.Empty{})
+	if err != nil {
+		return nil, err
+	}
+	var allPodEntries []*common.RegistrationEntry
+	nodeIDPrefix := fmt.Sprintf("%s/", r.RootID)
+
+	for _, maybePodEntry := range allEntries.Entries {
+		if strings.HasPrefix(maybePodEntry.ParentId, nodeIDPrefix) {
+			allPodEntries = append(allPodEntries, maybePodEntry)
+		}
+	}
+	return allPodEntries, nil
+}
+
+func (r *PodReconciler) getObject() ObjectWithMetadata {
+	return &corev1.Pod{}
+}
+
+func (r *PodReconciler) forEachPodSubsetEndpointAddress(subset corev1.EndpointSubset, action func(corev1.EndpointAddress)) {
+	for _, address := range subset.Addresses {
+		if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+			action(address)
+		}
+	}
+	for _, address := range subset.NotReadyAddresses {
+		if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+			action(address)
+		}
+	}
+}
+
+func (r *PodReconciler) forEachPodEndpointAddress(endpoints *corev1.Endpoints, thing func(corev1.EndpointAddress)) {
+	for _, subset := range endpoints.Subsets {
+		r.forEachPodSubsetEndpointAddress(subset, thing)
+	}
+}
+
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager, builder *ctrlBuilder.Builder) error {
+	if r.AddPodDNSNames {
+		builder.Watches(&source.Kind{Type: &corev1.Endpoints{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
```

Comment:
Why no endpoint reconciler?

Reply:
It felt cleaner to have the Pod controller completely own pod related entries, rather than spreading management of a single entry over multiple controllers. So the pod controller has all the code to find/create and mutate its entry, and reconciliation of the Pod needs to be triggered when Endpoints change.

I think there might actually be a bug here that maybe it wont be notified of matching Endpoints being deleted.